homeassistant/components/bryant_evolution/coordinator.py

```diff
@@ -0,0 +1,190 @@
+class EvolutionCoordinator(DataUpdateCoordinator[EvolutionState]):
+    """Class to manage fetching data from the Evolution API on a particular tty."""
+
+    # Note: directly accessing the data member of this class is discouraged.
+    # Prefer using the provided read_ functions, which are less error-prone.
+
+    def __init__(
+        self, hass: HomeAssistant, tty: str, system_zones: list[tuple[int, int]]
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            logger=_LOGGER,
+            name=f"Bryant Evolution at {tty}",
+            update_interval=timedelta(minutes=1),
+            update_method=lambda: _read_evolution_state(
+                self.system_zone_tuples, self.tty
+            ),
+            always_update=False,
+        )
+        self.system_zone_tuples = system_zones
+        self.tty = tty
+
+    def read_current_temperature(self, sys_id: int, zone_id: int) -> int:
+        """Return the current temperature."""
+        return self.data.systems[sys_id].zones[zone_id].curr_temp
+
+    def read_fan_mode(self, sys_id: int) -> str:
+        """Return the current fan mode."""
+        return self.data.systems[sys_id].fan_mode
+
+    def read_hvac_mode(self, sys_id: int) -> HVACMode:
+        """Return the current HVAC mode."""
+        evolution_mode = self.data.systems[sys_id].hvac_mode
+        mode_enum = {
+            "HEAT": HVACMode.HEAT,
+            "COOL": HVACMode.COOL,
+            "AUTO": HVACMode.HEAT_COOL,
+            "OFF": HVACMode.OFF,
+        }.get(evolution_mode.upper())
+        if mode_enum is None:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="failed_to_parse_hvac_mode",
+                translation_placeholders={"mode": evolution_mode},
+            )
+        return mode_enum
+
+    def read_target_temperatures(
+        self, sys_id: int, zone_id: int
+    ) -> tuple[int | None, int | None, int | None]:
+        """Return the target temperatures.
+
+        Specifically, returns a tuple with three integers. It will be either:
+        1. (target-temp, None, None) if the mode is HEAT or COOL
+        2. (None, heating-setpoint, cooling-setpoint) if the mode is HEAT_COOL.
+        """
+        mode = self.read_hvac_mode(sys_id)
+        # Return target_temperature or target_temperature_{high, low} based on mode.
+        match mode:
+            case HVACMode.HEAT:
+                return (self.data.systems[sys_id].zones[zone_id].htsp, None, None)
+            case HVACMode.COOL:
+                return (self.data.systems[sys_id].zones[zone_id].clsp, None, None)
+            case HVACMode.HEAT_COOL:
+                htsp = self.data.systems[sys_id].zones[zone_id].htsp
+                clsp = self.data.systems[sys_id].zones[zone_id].clsp
+                return (None, htsp, clsp)
+            case HVACMode.OFF:
+                return (None, None, None)
+            case _:
```

Comment:
Not a fan of this function returning a tuple. A helper function should only do one thing. And if you split this up, the only thing that I would put in a function is for target_temperature. target_temperature_low and target_temperature_high can be covered with a one-liner in climate.py (`self._attr_target_temperature_high = htsp if mode is HVACMode.HEAT_COOL else None`), it adds unnecessary complexity having a function for one-liners.

And I am also not a fan, that all these helper functions are methods of the coordinator. Move them maybe to a separate module, a good name would be helpers.py or util.py and import in climaty.py 

`self._attr_target_temperature = target_temperature(self.zone, self._attr_hvac_mode)`

```
   def target_temperature(zone: _ZoneState, mode: HVACMode | None) -> int | None:
    """."""

    if mode is HVACMode.HEAT:
        return zone.htsp
    if mode is HVACMode.COOL:
        return zone.clsp
    return None

```

Reply:
so, I agree that returning a tuple is a bit ick; I'm happy to make it return a TargetTemperatures dataclass with three fields (tgt, tgt_low, tgt_high) if you'd like.

I'd prefer to keep one function that returns three temperatures, rather than splitting this as you suggest.

The reason is that there are invariants across the three fields: the low/high setpoints are both set if and only if the target setpoint is not set, and this is controlled by whether or not we're in mode HEAT_COOL. (This is just how the climate entity thinks about target temperatures; we can't change it.)

The current code makes all of these invariants very clear in a single function: a reader will see exactly what's going on in a few lines of code. If we split the code in the way you suggest, we would have if statements around setting _low and _high in the climate entity (htsp if mode is heat_cool else none) and if statements here. Also, target_temperature as you define it introduces an additional return value (none) if called in an "invalid" state (not mode heat or cool) which the current code doesn't have.